datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,332 @@
+import os
+from typing import Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkMetadata,
+    DocumentMetadataFilter,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+import json
+import psycopg
+from psycopg import sql
+from psycopg_pool import ConnectionPool
+import logging
+from services.date import to_unix_timestamp
+
+# Read environment variables for Postgres
+
+POSTGRES_HOST = os.environ.get("POSTGRES_HOST", "localhost")
+POSTGRES_PORT = os.environ.get("POSTGRES_PORT", "5433")
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_EMBEDDING_INDEX = os.environ.get(
+    "POSTGRES_EMBEDDING_INDEX", "chatgpt_datastore_embedding_index"
+)
+POSTGRES_DOCID_INDEX = os.environ.get(
+    "POSTGRES_EMBEDDING_INDEX", "chatgpt_datastore_docid_index"
+)
+POSTGRES_UPSERT_COMMAND = os.environ.get("POSTGRES_UPSERT_COMMAND", "INSERT")
+POSTGRES_MIN_ROWS_FOR_INDEX = int(os.environ.get("POSTGRES_MIN_ROWS_FOR_INDEX", 1000))
+POSTGRES_MIN_NEW_ROWS_FOR_REINDEX = int(
+    os.environ.get("POSTGRES_MIN_NEW_ROWS_FOR_REINDEX", 1000)
+)
+
+assert POSTGRES_UPSERT_COMMAND in ["COPY", "INSERT"]
+
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        conninfo = (
+            "postgresql://"
+            + POSTGRES_USERNAME
+            + ":"
+            + POSTGRES_PASSWORD
+            + "@"
+            + POSTGRES_HOST
+            + ":"
+            + POSTGRES_PORT
+            + "/"
+            + POSTGRES_DATABASE
+        )
+        self.pool = ConnectionPool(conninfo)
+        conn = self.pool.getconn()
+        conn.autocommit = True
+        # Insert the vector and text into the database
+        logging.info("Creating table %s" % POSTGRES_TABLENAME)
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        index_exists_staetment = (
+            "SELECT EXISTS (SELECT 1 FROM pg_indexes  WHERE tablename = '%s' AND indexname = '%s')"
+            % (POSTGRES_TABLENAME, POSTGRES_DOCID_INDEX)
+        )
+        cur.execute(index_exists_staetment)
+        self.docid_index_exists = cur.fetchall()[0][0]
+
+        if not self.docid_index_exists:
+            index_doc_ids = (
+                "CREATE INDEX CONCURRENTLY ON %s (doc_id) " % POSTGRES_TABLENAME
```

Comment:
doc_id should be a unique index and it should include chunk_id to enforce upsert semantics with `ON DUPLICATE KEY UPDATE`, and we can skip the separate pre-existence check, by including that in the creation statement.
```suggestion
                "CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS ON %s (doc_id, chunk_id) " % POSTGRES_TABLENAME
```